File: samples/CameraAccessAndroid/app/src/main/java/com/meta/wearable/dat/externalsampleapps/cameraaccess/display/WidgetBoardView.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 * All rights reserved.
 *
 * This source code is licensed under the license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.meta.wearable.dat.externalsampleapps.cameraaccess.display

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp

@Composable
fun WidgetBoardView(
    widgets: List<WidgetSpec>,
    modifier: Modifier = Modifier,
) {
    Column(modifier = modifier, verticalArrangement = androidx.compose.foundation.layout.Arrangement.spacedBy(12.dp)) {
        for (widget in widgets) {
            WidgetCard(widget = widget)
        }
    }
}

@Composable
private fun WidgetCard(widget: WidgetSpec) {
    when (val kind = widget.kind) {
        is WidgetSpec.Kind.Text -> TextWidgetCard(title = kind.title, bodyText = kind.body)
        is WidgetSpec.Kind.Image ->
            ImageWidgetCard(
                imageUrl = kind.imageKind.url,
                caption = kind.caption?.takeIf { it.isNotEmpty() } ?: kind.imageKind.defaultCaption,
            )
        is WidgetSpec.Kind.Table ->
            TableWidgetCard(title = kind.title, columns = kind.columns, rows = kind.rows)
    }
}

@Composable
private fun TextWidgetCard(title: String?, bodyText: String) {
    Column(
        modifier = Modifier.fillMaxWidth().displayCardStyle().padding(18.dp),
    ) {
        if (!title.isNullOrEmpty()) {
            Text(
                text = title,
                fontSize = 18.sp,
                fontWeight = FontWeight.Bold,
                color = Color.White,
            )
        }
        if (bodyText.isNotEmpty()) {
            Text(
                text = bodyText,
                fontSize = 15.sp,
                color = Color.White.copy(alpha = 0.85f),
            )
        }
    }
}

@Composable
private fun ImageWidgetCard(imageUrl: String, caption: String) {
    Column(
        modifier = Modifier.fillMaxWidth().displayCardStyle().padding(16.dp),
    ) {
        RemoteImage(
            url = imageUrl,
            modifier =
                Modifier.fillMaxWidth()
                    .height(150.dp)
                    .clip(RoundedCornerShape(14.dp)),
        )
        Text(
            text = caption,
            fontSize = 13.sp,
            color = Color.White.copy(alpha = 0.6f),
            modifier = Modifier.padding(top = 8.dp),
        )
    }
}

@Composable
private fun TableWidgetCard(
    title: String?,
    columns: List<String>?,
    rows: List<List<String>>,
) {
    val columnCount = maxOf(columns?.size ?: 0, rows.maxOfOrNull { it.size } ?: 0)

    Column(
        modifier = Modifier.fillMaxWidth().displayCardStyle().padding(20.dp),
    ) {
        if (!title.isNullOrEmpty()) {
            Text(
                text = title,
                fontSize = 20.sp,
                fontWeight = FontWeight.Bold,
                color = Color.White,
            )
        }
        if (!columns.isNullOrEmpty()) {
            TableRow(cells = columns, columnCount = columnCount, isHeader = true)
        }
        for (row in rows) {
            TableRow(cells = row, columnCount = columnCount, isHeader = false)
        }
    }
}

@Composable
private fun TableRow(cells: List<String>, columnCount: Int, isHeader: Boolean) {
    Row(modifier = Modifier.fillMaxWidth().padding(top = 4.dp)) {
        for (i in 0 until maxOf(columnCount, 1)) {
            Text(
                text = cells.getOrNull(i).orEmpty(),
                modifier = Modifier.weight(1f),
                fontSize = if (isHeader) 12.sp else 15.sp,
                color = if (isHeader) Color.White.copy(alpha = 0.6f) else Color.White,
            )
        }
    }
}

@Composable
fun DisplayPreviewBoard(
    widgets: List<WidgetSpec>,
    modifier: Modifier = Modifier,
) {
    // Android SDK display content is builder-only; mirror the on-glasses layout natively.
    WidgetBoardView(widgets = widgets, modifier = modifier)
}

@Composable
private fun Modifier.displayCardStyle(): Modifier {
    val shape = RoundedCornerShape(18.dp)
    return clip(shape)
        .background(Color.Black.copy(alpha = 0.55f), shape)
        .border(1.dp, Color.White.copy(alpha = 0.12f), shape)
}

File: samples/CameraAccess/CameraAccess/Display/WidgetBoardView.swift
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 * All rights reserved.
 *
 * This source code is licensed under the license found in the
 * LICENSE file in the root directory of this source tree.
 */

//
// WidgetBoardView.swift
//
// Native SwiftUI renderer for a dynamic list of widgets (WidgetSpec) produced from
// a Gemini `render_widgets` tool call. Renders the three widget types — text, image,
// and table — each as a card in the shared Display card style.
//

import SwiftUI

struct WidgetBoardView: View {
  let widgets: [WidgetSpec]

  var body: some View {
    VStack(spacing: 12) {
      ForEach(widgets) { widget in
        WidgetCard(widget: widget)
      }
    }
  }
}

private struct WidgetCard: View {
  let widget: WidgetSpec

  var body: some View {
    switch widget.kind {
    case let .text(title, content):
      TextWidgetCard(title: title, bodyText: content)
    case let .image(url, caption):
      NativeImageWidget(
        imageURL: URL(string: url),
        caption: (caption?.isEmpty == false) ? caption! : "Image"
      )
    case let .table(title, columns, rows):
      DynamicTableWidget(title: title, columns: columns, rows: rows)
    }
  }
}

// MARK: - Text

private struct TextWidgetCard: View {
  let title: String?
  let bodyText: String

  var body: some View {
    VStack(alignment: .leading, spacing: 6) {
      if let title, !title.isEmpty {
        Text(title)
          .font(.system(size: 18, weight: .bold))
          .foregroundColor(.white)
      }
      if !bodyText.isEmpty {
        Text(bodyText)
          .font(.system(size: 15))
          .foregroundColor(.white.opacity(0.85))
          .fixedSize(horizontal: false, vertical: true)
      }
    }
    .frame(maxWidth: .infinity, alignment: .leading)
    .padding(18)
    .displayCardStyle()
  }
}

// MARK: - Table

private struct DynamicTableWidget: View {
  let title: String?
  let columns: [String]?
  let rows: [[String]]

  private var columnCount: Int {
    max(columns?.count ?? 0, rows.map(\.count).max() ?? 0)
  }

  var body: some View {
    VStack(alignment: .leading, spacing: 8) {
      if let title, !title.isEmpty {
        Text(title)
          .font(.system(size: 20, weight: .bold))
          .foregroundColor(.white)
      }
      if let columns, !columns.isEmpty {
        row(cells: columns, isHeader: true)
      }
      ForEach(Array(rows.enumerated()), id: \.offset) { item in
        row(cells: item.element, isHeader: false)
      }
    }
    .frame(maxWidth: .infinity, alignment: .leading)
    .padding(20)
    .displayCardStyle()
  }

  private func row(cells: [String], isHeader: Bool) -> some View {
    HStack(alignment: .top, spacing: 12) {
      ForEach(0 ..< max(columnCount, 1), id: \.self) { i in
        Text(i < cells.count ? cells[i] : "")
          .frame(maxWidth: .infinity, alignment: .leading)
      }
    }
    .font(.system(size: isHeader ? 12 : 15))
    .foregroundColor(isHeader ? .white.opacity(0.6) : .white)
  }
}

#Preview("Widget board") {
  ZStack {
    LinearGradient(colors: [Color(white: 0.4), Color(white: 0.1)], startPoint: .top, endPoint: .bottom)
      .ignoresSafeArea()
    ScrollView {
      WidgetBoardView(widgets: [
        WidgetSpec(kind: .text(title: "Vienna afternoon", body: "Here's a quick plan for the old town.")),
        WidgetSpec(kind: .image(url: WidgetImageKind.map.url, caption: "Hofburg area")),
        WidgetSpec(kind: .table(
          title: "Schedule",
          columns: ["Time", "Stop"],
          rows: [["09:00", "Hofburg"], ["10:30", "Stephansdom"], ["12:00", "Belvedere"]]
        )),
      ])
      .padding()
    }
  }
}
